x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportSetUpgradeModeAction.java

```diff
@@ -0,0 +1,327 @@
+     * <p>
+     * If we do not force an allocation change for the Datafeed tasks, they will never start again, since they were isolated.
+     * <p>
+     * Datafeed tasks keep the state as `started` and Jobs stay `opened`
+     *
+     * @param tasksCustomMetaData Current state of persistent tasks
+     * @param listener            Alerted when tasks are unassignd
+     */
+    private void unassignPersistentTasks(PersistentTasksCustomMetaData tasksCustomMetaData,
+                                         ActionListener<List<PersistentTask<?>>> listener) {
+        List<PersistentTask<?>> datafeedAndJobTasks = tasksCustomMetaData
+            .tasks()
+            .stream()
+            .filter(persistentTask -> (persistentTask.getTaskName().equals(MlTasks.JOB_TASK_NAME) ||
+                persistentTask.getTaskName().equals(MlTasks.DATAFEED_TASK_NAME)))
+            .collect(Collectors.toList());
+
+        TypedChainTaskExecutor<PersistentTask<?>> chainTaskExecutor =
+            new TypedChainTaskExecutor<>(client.threadPool().executor(executor()), r -> true, ex -> true);
+
+        for (PersistentTask<?> task : datafeedAndJobTasks) {
+            chainTaskExecutor.add(
+                chainedTask -> persistentTasksClusterService.unassignPersistentTask(task.getId(),
+                    task.getAllocationId(),
+                    AWAITING_UPGRADE.getExplanation(),
+                    chainedTask)
+            );
+        }
+        chainTaskExecutor.execute(listener);
+    }
+
+    private void isolateDatafeeds(PersistentTasksCustomMetaData tasksCustomMetaData,
+                                  ActionListener<List<IsolateDatafeedAction.Response>> listener) {
+        Set<String> datafeedsToIsolate = MlTasks.startedDatafeedIds(tasksCustomMetaData);
+
+        TypedChainTaskExecutor<IsolateDatafeedAction.Response> isolateDatafeedsExecutor =
+            new TypedChainTaskExecutor<>(client.threadPool().executor(executor()), r -> true, ex -> true);
+
+        datafeedsToIsolate.forEach(datafeedId -> {
+            IsolateDatafeedAction.Request isolationRequest = new IsolateDatafeedAction.Request(datafeedId);
+            isolateDatafeedsExecutor.add(isolateListener ->
+                executeAsyncWithOrigin(client, ML_ORIGIN, IsolateDatafeedAction.INSTANCE, isolationRequest, isolateListener)
+            );
+        });
+
+        isolateDatafeedsExecutor.execute(listener);
+    }
+
+    private void awaitCondition(CheckedSupplier<Boolean, Exception> condition,
+                                String reason,
+                                TimeValue timeout,
+                                ActionListener<AcknowledgedResponse> listener) {
+        threadPool.generic().execute(() -> {
+            try {
+                long msWaited = 0;
+                long tick = 1;
+                boolean done = condition.get();
+                while (msWaited <= timeout.millis() && done == false) {
+                    Thread.sleep(tick);
```

Comment:
This is tying up a thread while it waits.  If "sleeping" is the only way to wait for the required conditions then it should instead schedule a timer to run the check after a wait time.

It may be possible to avoid sleeping altogether though.  For the restarts I think we should just wait for the task assignments to be changed from `AWAITING_UPGRADE` as I've said in other comments, and that can be achieved using `PersistentTasksService.waitForPersistentTaskCondition`.  The implementation of this is using a cluster state listener.

Maybe there's also a way to wait for (non-persistent) task status using a cluster state listener.  We should check.